### PR TITLE
Make nominating more controllable

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
This PR contains code split off from the rest of the work that's in progress for https://github.com/bpfkorea/agora/issues/1004

These changes look nice on their own and they can be merged before the rest of the work is submitted.

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
Fixed commit message.

File: source/agora/node/Ledger.d [hunk 1]
```diff
@@ -340,44 +340,38 @@ public class Ledger
...
-        if (this.pool.length < Block.TxsInBlock)
-            return;
-
+        data = ConsensusData.init;
         const next_height = Height(this.getBlockHeight() + 1);
         auto utxo_finder = this.utxo_set.getUTXOFinder();
 
         this.enroll_man.getEnrollments(data.enrolls,
             Height(this.getBlockHeight()));
+
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: Whitespace change.

File: source/agora/node/Ledger.d [hunk 2]
```diff
@@ -527,14 +521,17 @@ public class Ledger
     }
 }
 
+/// Note: these unittests historically assume a block always contains
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: This comment belongs in a future commit, not here.

File: source/agora/test/Base.d [hunk 3]
```diff
@@ -351,6 +352,18 @@ public class FakeClockBanManager : BanManager
     public override void dump () { }
 }
 
+/// Nominator with custom rules for when blocks should be nominated
+public extern (C++) class TestNominator : Nominator
+{
+extern(D):
+
```
Comment (AndrejMitrovic) [inline source/agora/test/Base.d]: ///

File: source/agora/test/Base.d [hunk 4]
```diff
@@ -853,6 +866,13 @@ public class TestValidatorNode : Validator, TestAPI
     {
         return this.qc;
     }
+
+    /// Returns an instance of a TestNominator with custom nominating behavior
```
Comment (AndrejMitrovic) [inline source/agora/test/Base.d]: custom -> customizable

File: source/agora/test/Base.d [hunk 5]
```diff
@@ -356,12 +357,37 @@ public class FakeClockBanManager : BanManager
 public extern (C++) class TestNominator : Nominator
 {
 extern(D):
+    /// number of txs required for nomination
+    private ulong txs_to_nominate;
```
Comment (AndrejMitrovic) [inline source/agora/test/Base.d]: This doesn't belong in this PR.

**Comment by AndrejMitrovic** [comment]
Ok resolved comments and removed code which didn't belong here.

File: source/agora/test/Base.d [hunk 6]
```diff
@@ -853,6 +866,13 @@ public class TestValidatorNode : Validator, TestAPI
     {
         return this.qc;
     }
+
+    /// Returns an instance of a TestNominator with customizable behavior
+    protected override TestNominator getNominator ( NetworkManager network,
```
Comment (linked0) [inline source/agora/test/Base.d:871]: This function is overridden so the signature should be the same as that of the superclass. But I know what you intended.

**Review by linked0 — COMMENTED** [review]
Very good refactoring, but I have some comments.

File: source/agora/test/EnrollmentManager.d [hunk 7]
```diff
@@ -376,7 +356,7 @@ unittest
         }
 
         ///
-        protected override Nominator getNominator (
+        protected override TestNominator getNominator (
```
Comment (linked0) [inline source/agora/test/EnrollmentManager.d:359]: Same as the above comment.

File: source/agora/test/Base.d (hunk 6 above)
Comment (AndrejMitrovic) [reply source/agora/test/Base.d:871]: https://en.wikipedia.org/wiki/Covariant_return_type

Comment (AndrejMitrovic) [reply source/agora/test/Base.d:871]: There is also a very good reason to do this. There is an override of this method here: https://github.com/bpfkorea/agora/blob/97bfb80edad10731d0a5f41647428913a51dbafb/source/agora/test/EnrollmentManager.d#L383

If I didn't use `TestNominator` in `TestValidatorNode` then the code in `EnrollmentManager` could be written as:

```D
static class BadNominator : Nominator  // NOT TestNominator
{
    ...
}

protected override Nominator getNominator ( ... ) { return new BadNominator(...); }
```

This means it would circumvent the TestNominator class. But we want all derived classes to be forced to instantiate either `TestNominator` or a derived class from `TestNominator`. Changing the return type to be narrower (TestNominator in this PR line) will force the compiler to check this. It will be a compile-time error if the derived class returns an instance of a Nominator and not a TestNominator.

**Comment by AndrejMitrovic** [comment]
@linked0 can you approve / merge this? It will unblock the rest of my work.

**Review by linked0 — APPROVED** [review]

**Comment by AndrejMitrovic** [comment]
Thank you!
